### ISSUE TITLE
Water Simulation?

### ISSUE THREAD

**Issue by iUltimateLP** [AndrewScheidecker/BrickGame#20]
Okay, lets get to a real hot topic. Water simulation? How would I proceed? And, how do I generate just a simple ocean? I think I need to drop my project because this is all stuff I'm not aware of :cry:

**Comment by miguelemosreverte** [comment]
Hey I did it. Not the real-time flooding part, but the part where water instantly floods what it can, being the only parameter to it that it should never flood anything taller than its origin point. (Which is a gross simplification, given pascal law and all).

So, the idea is simple, for it to work fast I developed a data structure (not compressed... which is bad), that stores every possible 2d surface for a given 3d lake/ocean.

I can make a pull request if you guys think it is a good idea. Oh, but here is the kicker, and this Scheidecker probably knew all the time: About the water. You can drag and drop any water material to the surface material of a given block material. Which is awesome. Here is a screenshoot of my results: (You can find the water material I use here: https://github.com/UE4-OceanProject/OceanProject)

screenshoots: 

![water](https://cloud.githubusercontent.com/assets/9152392/13815785/ad33df2a-eb6a-11e5-9939-7ad7c8ffb696.png)

And by doing some tweaking now the blocks under water blocks are rendered too:

![translucent water with the blocks rendered correctly](https://cloud.githubusercontent.com/assets/9152392/13817109/bff8873c-eb6f-11e5-800d-9f6be0f84d1f.png)
![under water surface](https://cloud.githubusercontent.com/assets/9152392/13817110/bffd2d78-eb6f-11e5-89b4-b0f3c02ac031.png)

**Comment by iUltimateLP** [comment]
That looks awesome!
Sure thing that you can apply any material you want onto stuff, so I would like a pull request :+1: :joy:

**Comment by miguelemosreverte** [comment]
Give me a sec (maybe a couple days) to clean it up so it is ready to merge, and then I will make the pull request. About the data structure, on a basic level it stores the FInt3 Coordinates of the bricks that conform each one of the XY slices of a XYZ lake. Something I am working on is to reuse that information for custom pathfinding. Because if you use the collision of BrickGrid along with UE4 NavMesh you will get unrealistic movement: People will find no trouble climbing unnecesary elevations. But if you use the data of the lakes, then you could develop your own pathfinding (which could be cheaper) where the people will always walk on flat surfaces, unless its not possible.

**Comment by iUltimateLP** [comment]
Nice to hear that somebody is actually working here :+1: Looking forward :)

**Comment by miguelemosreverte** [comment]
My goal is an RTS on top of a voxel world, whats yours?

**Comment by iUltimateLP** [comment]
A crossover between Starbound + Minecraft + Factorio :joy: Not even sure if I bring this to a certain point

**Comment by miguelemosreverte** [comment]
Sounds like you are going to work a lot more than I on the gameplay dynamics. Factorio especially, is all about that. Tons of different options and winning strategies.

**Comment by iUltimateLP** [comment]
Yep, and I'm not sure if I even finish this, but if I do, it'll go on Steam :)

**Comment by miguelemosreverte** [comment]
Greenlight is the way to go. Kickstarter + Greenlight, really. Kickstarter can be used for the money, but the exposure is in my opinion more important. There you can find artists ready to work on the art on top of your game. Maybe not kickstarter, is not like I have done proper research, but on a general scale, that would be the way to go. Develop the logic, expose the product, find the team (ussually the artists), and then Greenlight. Maybe I am wrong but, again, its not like I am close to that part of the developing process so.

What I sometimes find weird is what is the motivation of Andrew to build this absolutly fantastic piece of code. I mean, you want to make your game, I want to make mine, whats up with the creator of the     most complicated part of the work we are working on top of? I would not be suprised if tomorrow a game is shipped and it was made by Andrew. I am ust curious what genre would it be of. One thing is certain, it would have a Minecraft feel to it for sure.

**Comment by iUltimateLP** [comment]
I don't think he is working on this currently. I hate people calling games "Minecraft Ripoffs" just because they share the way of doing a terrain and a building system...

**Comment by miguelemosreverte** [comment]
Indeed! Is like minecraft has the monopoly of voxel use in games! That said, given the popularity of it, it can be said that games using cubes as the representation of voxels have a similar aestethic.

**Comment by iUltimateLP** [comment]
Does your simulation also generate water lakes through the world or does it only support "oceans" yet?

**Comment by miguelemosreverte** [comment]
Not really a simulation, but a reduntant database. The information I am creating is implicit on the Region.BrickContents. Is satifies the following query:

What bricks can I "flood" from this location. Where "flood" means access all adjacent empty bricks of the same height or less.

You could access this information by using a search algorithm every time. Or you could "cook" it. Which bring us to the redundant database.
Now, about its data structure:

It works like this. Inside the function SetBrick it checks if the brick set is of a given material. Lets say the material that floods depressions is 1. And the material used for the water effect is 9.

Then it accesses with the given coordinate to an index file that says what lake this block belongs to. On this file, absolutly all the blocks coordinates are linked to LakeIndexes. Non Empty bricks has a LakeIndex of -1, as they cannot be flooded. A LakeIndex is the key to then access the Lake information itself. Right now if you see a LakeIndex of 4, it means the LakeInitialCoordinate was X = 0, Y= 0, and Z = 4.

Anyway. With the LakeIndex found, now you can access the lake information, which is inside the file that has the LakeIndex as name. 

Inside the file you can access all the floodable bricks, and for each one of them you change their material and boom. The lake/ocean is flooded.

**Comment by AndrewScheidecker** [comment]
@miguelemosreverte, that looks great!

I'd definitely merge a pull request to make translucent bricks render correctly. If you submit a PR for water flow, I'd consider merging it, but only if I was prepared to support the code (though I know my support lately has been lacking). Even if I don't merge it, the PR can still be helpful for other people who want to use your changes.

From reading your overview of how it works, my primary concern would be the performance of recomputing the flood volumes when a brick changes. If you add or remove a single brick, it could cause the flooding information to change over a very large area, so it might be hard to do incrementally. I think that's an argument for something more like Minecraft, Dwarf Fortress, and other games that implement water that flows from block to block in real-time: it localizes the effect of modifying the bricks.

> What I sometimes find weird is what is the motivation of Andrew to build this absolutly fantastic piece of code. I mean, you want to make your game, I want to make mine, whats up with the creator of the most complicated part of the work we are working on top of? I would not be suprised if tomorrow a game is shipped and it was made by Andrew.

BrickGame was a fun side-project I did when UE4 was first released publically, and wanted to provide an example to folks how to implement rendering plugins like this. I've just been doing minimal maintenance work since then, except for my VXGI experiment. You don't have to worry about me competing with you if you want to release a game using this code!

**Comment by miguelemosreverte** [comment]
I knew you where going to be specially attracted to the implementation of translucent bricks, because while flooding algorithms are cool and all, they can made with time and effort. But tweaking the render code? That requires more thought! And you specially would be, if not the only one, the one to know this since you are the creator of it! 

So, yeah. With that to a side (looking back to it I just added two or three lines of code, its not impresive to look at if you do not know the complexity of the code), about the flooding algorithm:

Cellular automaton is a naive algorithm, it does not provide scalability. What I am looking for is to make the information much more redundant, to the point where PC A can say to PC B:

"PC A : Hey, remember the LakeSlice 204534 of the Region 300?
PC B: Yup.
PC A: Player 0 just flooded it. These are the holes done, so you know the flux.
PC B: Right. So given all the previously recorded data, we both know it will take an hour to fill.
PC A: Yup. Important locations flooded include: This guy castle, and etc.
PC B: I know, why are you telling me this? 
PC A: Sorry, sometimes I forget the flooding logic is deterministic.
PC B: Duh."

So, the idea would be that by having a reduntant dabatase communication would be small, if at all needed. About the real-time super cool flooding effects I believe a centralized cloud-based computing could solve the problem. It would give the clients the vertices of the poligon used for the wall of incoming water. Okay, I am rammbling away. I expect to have the Pull request ready for today .

If my database teacher heard me talking about redundancies as a good thing he would be so mad...

**Comment by iUltimateLP** [comment]
When I was thinking about basic water simulation, my idea was pretty simple: When you populate the blocks according to the noise (the three loops), you basicially get every block which is below a limit (like the material selection works right now) - the sea limit. Then when a block is under the sea limit, you create an additional block which goes some values below the original one in the height axis. That forms the ground of the sea. The block you had originally is the new sea block, so that gets the ocean shader on it. Two problems still: aligning the post process volume to fit under the blocks, and second: the transition between non-water and water would be rough, as non-existent.

**Comment by miguelemosreverte** [comment]
Oh do not worry about the post-process volume. Andrew gave us the example on how to create redundant blocks on top of each other: You have the ones you use to render, and all is good. But now what about collision?? Well, as seen in the example given, you create collision boxes in the exact same spot. Do the same with post process and then with Physics Volumes (to add buoyancy forces, for example) and you are done!

I believe I should appoint this as my next task now that I see the urgency of it.
Now a problem I should solve too is when you use a tesselated material on the rendering brick, and the brick surface goes higher in space, but the collision, PP, and physics boxes are still the same size. Now thats a problem. And if Andrew does not hace a space under his slive then we are all doomed! hahaha

**Comment by iUltimateLP** [comment]
I look forward to test your pull request out! Thanks for contributing to this and helping me with it the same time!

**Comment by miguelemosreverte** [comment]
I guess so, but really, thanks to Andrew. He is exactly what every developer community needs. To code for fun speaks volume of his qualities as a programmer. Is like those olimpic runners that go to "work out" every day and by working out I mean do stuff we mortals can only imagine because we grow tired half way.

**Comment by miguelemosreverte** [comment]
ISP change complicated downloading the repository. Downloading it now. Delay 1 day.

**Comment by iUltimateLP** [comment]
Oh nice! Could this also allow glass blocks?

**Comment by miguelemosreverte** [comment]
sure. A list of transclucent materials could be implemented so that instead of just checking for material 9 it would check for the ones you added as translucent.

**Comment by iUltimateLP** [comment]
Nice stuff! I will test it out soon!

**Comment by miguelemosreverte** [comment]
First problem found: All this time it worked because the water flooded until it found a non empty cube.
But now, by spawning a translucent block with empty neighboors at the sides, this happens:

![image](https://cloud.githubusercontent.com/assets/9152392/13878570/184b6814-ecf1-11e5-8d2c-620f565df176.png)

The translucent cube is using the vertices of the floor correctly, if at all. This way the only well rendered translucent cubes are the ones surrounded by other cubes: (which had been the case until not long ago):

![image](https://cloud.githubusercontent.com/assets/9152392/13878732/044d9c82-ecf2-11e5-922e-07c427ab294d.png)

**Comment by iUltimateLP** [comment]
> It should be easy to fix.

For you my friend, for you :joy: :+1:

**Comment by miguelemosreverte** [comment]
I am trying to find a pattern:

Before
![image](https://cloud.githubusercontent.com/assets/9152392/13883546/a7f6881a-ed08-11e5-870a-54c7f9ace7c3.png)
After
![image](https://cloud.githubusercontent.com/assets/9152392/13883573/cf475c96-ed08-11e5-9311-e88f03f57e03.png)

**Comment by iUltimateLP** [comment]
it looks like there is an vertex missing to make it a rectangle..

**Comment by miguelemosreverte** [comment]
Yup. 
![image](https://cloud.githubusercontent.com/assets/9152392/13884470/ecdef1ca-ed0c-11e5-932e-4624744e2d07.png)

**Comment by iUltimateLP** [comment]
Couldn't you just add another vertex to that position (maybe there are two missing because of the lower one of the two)?

**Comment by miguelemosreverte** [comment]
Exactly. Trouble is to understand why this has gone wrong, since in theory this should not be happening. But... oh well. It will be a matter of looking at the code for a while sipping coffee. Will update  inmediatly when the culprit is found.

**Comment by miguelemosreverte** [comment]
Bug free version uploaded to the pull request:
![image](https://cloud.githubusercontent.com/assets/9152392/13887834/acb35dd8-ed1c-11e5-9183-2d9b25fa0530.png)

**Comment by iUltimateLP** [comment]
Good work!!

**Comment by iUltimateLP** [comment]
My little test:
![image](https://cloud.githubusercontent.com/assets/8491339/13898477/649748e6-edd2-11e5-90a5-af755640ca0e.png)
So, back to water. I'll test out later if I can just put a water shader on a block and place it that way. In generation terms: Would you mind sharing the generation of the ocean? Maybe in a seperate Pull request?

**Comment by miguelemosreverte** [comment]
Absolutly. But first I think I should share the database first. Because it was the implementation of it that speeded up the access and modification of bricks contents.

**Comment by iUltimateLP** [comment]
Hmh, I don't get where "having different brickgrids in one level and allowing them to join each other" and "water flooding" meets. Oh, and could you please do a new post if you find something out instead of edits, because I don't check this page daily and then I miss some stuff :joy:

**Comment by miguelemosreverte** [comment]
Yeah, I do not see that connection either. It derrailed there, but going back to where we left it, the water flooding algorithm needs a database to work. One could argue there is already one, the information stored on RAM, but I like the idea of using data stored on disk. Being one of the reasons that after you go though the trouble of saving the created blocks on disk you will not have to calculate their position again. Second argument is that once you create a flow of information, now you can use it to import information from outside BrickGame. Example given: Heighmaps created on software like WorldMachine.

And finally I know that there is implemented a way of saving the entire game state to disk. And I believe that my own custom implementation of a database is better because instead of reading and writing the entirety of the game state it just reads one region at the time. Oh well. I have a branch ready to PR once the first PR is accepted.

Good stuff about my custom database:

.It allows for importing of external data (like Minecraft schematics (is this legal?))
.On top of it a redundant database can be creating, (the one the flooding algorithm uses),
that stores 2d "floodable" surfaces. Surfaces which frontiers are made of non emtpy blocks.

This redundant database, (the second one on top of the first) would not only allow for the instant flooding explained before. But most importantly it would give a domain for pathfinding algorithms to work on. I have tested navigation on top of the bricks and the guys have no problem going up and down again and again. This way they would prefer to take longer paths, by walking horizontal.

In short: The first database that stores the Region.BrickContents is ready for PR, and its main attraction is that it allows for communication with external sources.

The second one is redundant of the first and adds a really cool extra layer of information.

**Comment by iUltimateLP** [comment]
I read stuff about saving the game here, and I actually did that before. Andrew's way of saving the game was just a Save Game Object containing the Brick Grid Data, the player transform and the daytime, and that save file was 600 MB big on my end, so I decided to make up a simple compressed save game system, which decreased the file size to 2MB. The code for that is [here.](https://gist.github.com/iUltimateLP/e38a134c4a674a12c1b4)
The overloaded << operator for FBrickRegion (BrickGrid/Classes/BrickGridComponent.h):
```
//Overload << for FBrickRegion
FORCEINLINE FArchive& operator<<(FArchive& Ar, FBrickRegion& data)
{
	Ar << data.BrickContents; //TArray<uint8>
	Ar << data.Coordinates; //Fint3
	Ar << data.MaxNonEmptyBrickRegionZs; //TArray<int8>

	return Ar;
}
```

**Comment by miguelemosreverte** [comment]
Hmm. Seems like competition! One thing is clear, both our ways use compression, and lots of it. You used the one packaged on FArchive, I used the UE4 library for zlib.

**Comment by iUltimateLP** [comment]
From what I read, you're using zlib directly where I use Unreal's wrappers for that, but in the end, it has the same result..

**Comment by miguelemosreverte** [comment]
Yup. Yours is cleaner.

**Comment by iUltimateLP** [comment]
I actually made it in a seperate plugin, but it could also go in the BrickGridComponent directly. Should I do a PR?

**Comment by miguelemosreverte** [comment]
Sure, it is good quality code and compression is needed. After that I will post my database PR, using your compression implementation instead of mine. Good stuff!

**Comment by iUltimateLP** [comment]
Should I keep it in a plugin or make it directly in the BrickGridComponent?

**Comment by miguelemosreverte** [comment]
My opinion is that you should do the later, since BrickGame is a plugin by itself.

**Comment by iUltimateLP** [comment]
I'm a bit new to the PR stuff.. Do I need to do my stuff on top of your code, when Andrew merges your PR or does the code get merged together automaticially..

**Comment by miguelemosreverte** [comment]
Uhm, I am kind of new too. But my idea would be to wait until my PR is accepted, then immediately do your own PR. Like, seconds later. hahaha.

**Comment by iUltimateLP** [comment]
```
	FString title, title2;
	title = Directory;
	title += FString::FromInt(RegionToSave.Coordinates.X);
	title += " ";
	title += FString::FromInt(RegionToSave.Coordinates.Y);
	title += " ";
	title += FString::FromInt(RegionToSave.Coordinates.Z);
	title += ".bin";
```    
You're saving each chunk inside a custom .bin file, like Minecraft does? Or is this just your method of saving, where you said mine is better?

**Comment by miguelemosreverte** [comment]
Yes, and when I said yours was cleaner I was talking about your implementation of compression:
Both our methods use compression. Yours use it in a more elegant way.

About my method of saving:
Yours, like Scheidecker's, save the entirety of the Regions on one big file.
Mine saves each Region on a different file. The reason for this is: When the moment of replicating information between PC's come, you wont want a big file. You will want small little files. So you can implement the little files later on, or do it like I did. To do it first, like I did, comes with the additional  advantage that you wont ever need to read the complete file every time you want to decompress it, which in the Scheidecker code happens every time you load the game. I believe F7 fires the loading event.

So, in the end it is a "Divide and Conquer" approach to the loading of the data. Once you analize it you realize it is the best way of loading the data, and using small chunks becomes the only reasonable way to do it. And that is why Minecraft happens to use a similar approach.

Talking about Minecraft: Is it legal to import Minecraft related data?

**Comment by iUltimateLP** [comment]
Yeah, I think your version is better..
To the Minecraft stuff: I think it is, because it's just a file format (an open one, see mcedit which is even open-source)..

**Comment by iUltimateLP** [comment]
So with this system we basicially load a save and this only loads 4 chunks or so in the game. If I enter the distance for loading another chunk, it loads that chunk from the file into the world?

**Comment by miguelemosreverte** [comment]
My own code works this way: On the Update function it checks is there is a file by the name of 0 0 0 for example. Then if there is one it loads it into memory, and if not then it can calculate the region and save it. I hope this answers your question, since I dont think I did understood it right.

**Comment by miguelemosreverte** [comment]
I have a separate python program to read schematics and then create uncompressed binary files with the contents ready to be used in BrickGame. (Once compressed).

**Comment by miguelemosreverte** [comment]
After doing some research I am sure you can do your pull request at the same time. Can wait for your PR and mine to be integrated with the project!

**Comment by iUltimateLP** [comment]
But my system saves the whole brick grid in one big file, whereas your system saves it in small files, so a PR from my side would be useless as I think..

**Comment by miguelemosreverte** [comment]
I would then use your compression implementation

**Comment by miguelemosreverte** [comment]
Right now I have some ugly function called CompressTo8bit or something along those lines. After your PR I would use your functions!

**Comment by iUltimateLP** [comment]
So we would go with the one-file-method again?

**Comment by miguelemosreverte** [comment]
No! Atleast not if I can do the third PR successfully.

First PR: Translucency
Second PR: (yours) Compression implemented in an elegant way
Third PR: (not yet confirmed) my own database of small little files
Fourth PR: (maybe?) my secondary database, redundant of the first
Fifth: water flooding algorithm

**Comment by miguelemosreverte** [comment]
*updated PR list

**Comment by iUltimateLP** [comment]
So If I do this right now, I work on top of a fresh clone of Andrew's repo or I take your Translucency branch for it?

**Comment by miguelemosreverte** [comment]
fresh one. Github should merge them later on with no problem.

**Comment by miguelemosreverte** [comment]
Github should manage to merge them later on with no problem.

**Comment by iUltimateLP** [comment]
PR is done!

**Comment by iUltimateLP** [comment]
Do you know how I could geht realistic reflections on my glass Material? The Screenshot I posted days ago was with a sphere reflection capture, which captured the sky only and not the Blocks, thats why it is shining so red.. Is there any method of capturing that at runtime / when cubes are generated?

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14022507/e66a7d44-f1be-11e5-8831-9ff83b84ba88.png)

SSR works fine (as good as it can), but you are right. The sphere reflection capture is not correctly implemented. I mean, it needs to recapture.

**Comment by miguelemosreverte** [comment]
Just some more pics.
![image](https://cloud.githubusercontent.com/assets/9152392/14023862/34adf5e8-f1c5-11e5-9ade-31b7f7232937.png)
![image](https://cloud.githubusercontent.com/assets/9152392/14023881/52250062-f1c5-11e5-88c8-24ad6e5adbec.png)

**Comment by miguelemosreverte** [comment]
Take a look at this specific time of the VXGI showreel made by Scheidecker:

https://youtu.be/_UYavLt0NpI?t=58

The wall is reflecting whats behind the player character. Maybe VXGI is the solution?

**Comment by iUltimateLP** [comment]
Yeah.. The thing is you can't force the ReflectionCapture to recapture at runtime.. Probably with C++ tho.. I'll find some time and dig into the code of that actor, because we actually just need to rebuild it when the terrain is generated, and it can't be too hard to find the function which is executed when you press the "Recapture" button.

VXGI is nice, but on my card (GTX 960 @ 4GB) it only gives me around 10 fps, and that is definitly not a solution for players..

**Comment by miguelemosreverte** [comment]
Ha! It gives me 0.5 to 5 fps on mine (Intel HD 4000) hahaha

But it looks awesome. About the digging, it sounds like the only way to do it. Cant wait for news! This would also make water look so much better...

**Comment by iUltimateLP** [comment]
Heh, theres a console command ```r.ReflectionCapture``` which rebuilds every reflection capture sphere in the map. I wasn't thinking that it's so easy! One note still: the other blocks (like rock) also have some kind of red shining, even underground, where they should be completely dark..

**Comment by miguelemosreverte** [comment]
Can confirm. 

![image](https://cloud.githubusercontent.com/assets/9152392/14045092/d205bbe0-f275-11e5-8ed4-9c4ffc81ab18.png)

About the underground lighting, (VXGI! VXGI! VXGI!) having the coordinates of the blocks surrounded by nothing (Moore's neighboors, not the von neumann's ones) would help: You would change the location of the sphere reflection capture when the player gets near.

No, forget about it: Check this out: By the looks of it, the bricks are not captured as reflections. Not sure about it, will test it multiplying the ammounts of sphere reflection captures and the ammount of bricks rendered, and then will update with the results.

![image](https://cloud.githubusercontent.com/assets/9152392/14046014/44fdd306-f27d-11e5-831c-7f6c9e79fbc5.png)

![image](https://cloud.githubusercontent.com/assets/9152392/14046035/683bc5a8-f27d-11e5-95df-8f0142868c79.png)

**Comment by iUltimateLP** [comment]
This could be due to the PostProcess, as one of them has a cubemap texture inside which looks like this empty sky you see. Anyway, how is it going with water stuft?

**Comment by miguelemosreverte** [comment]
Waiting for your PR to be accepted so that then I can submit my own about databases and redundant databases. But the water is ready to be used, as you know is just a material!

**Comment by iUltimateLP** [comment]
I mean the realtime flooding part :joy:

**Comment by miguelemosreverte** [comment]
Oh, the real time flooding? You see, thats exactly what I think we should not try to do. Personal computers cannot accept the burden of that. Never will. Centralized computing will always be the best approach to physics.

**Comment by iUltimateLP** [comment]
Hmh, but how do you want to keep it at least a bit realistic? If you place water on the ground it simply stays there?

**Comment by miguelemosreverte** [comment]
That mechanic would be forbidden for sure. The placing of water as a block. The only way you would interact with water would be by deforming the terrain. And the results would be absolutly realistic, what are you talking about? Think about it! The server does all the hard work!

**Comment by iUltimateLP** [comment]
Hm, I think in my game the player only puts a jar of water on the ground :joy:

**Comment by miguelemosreverte** [comment]
Then you should use what minecraft used

**Comment by iUltimateLP** [comment]
Would that be hard to do? Generally, the water source point is where the player placed the water. Then basicially, it checks for neighbours, and if there is no neighbour at a certain side, then I create another water block there (they need to have different heights, so thats where maybe some custom block creation code comes in (where the vertices are created)), and this process is done to all sides a specific amount of times (how long I want the water to spill away). Am I right with those ideas?

**Comment by miguelemosreverte** [comment]
Yup. Its called celullar automaton!

**Comment by miguelemosreverte** [comment]
About it being hard, it will be a great coding experience! You will get familiar with the C++ side of BrickGrid. By the way, the height difference, that should be implemented on the material. WorldPositionOffset would be the variable to use.

**Comment by iUltimateLP** [comment]
You're right, would be easier..

**Comment by miguelemosreverte** [comment]
Hello there, because I could not resist I did the minecraft water flooding, so check out my branch RealtimeFlooding!

.Right now it only expands horizontally
.Code is bad, needs refactoring
.The two coordinates output of the function WaterFlooding, used then in the function InvalidateChunkComponents, are not right. It works for a start, but needs fixing.

**Comment by iUltimateLP** [comment]
That looks awesome! If you now do it in the Z direction also and make blocks with different heights, you're my hero 😂

**Comment by miguelemosreverte** [comment]
https://www.youtube.com/watch?v=_JUTxvoXz7k

**Comment by miguelemosreverte** [comment]
i am trying to find that out right now. The algorithm is fine, bassically it says if empty downwards go down, if not then expand to the sides. The bug happens when a brick that is on the list of the ones that have water is on top of recently flooded bricks. Then it acts like this bricks are the ground and it expands too. I will solve it by saying it only can expand to the sides if the block underneath is not water. Not without permission, that is. Later on I would have to work on that. So that once the water touches ground it expands, and later on it goes higher in height and expands again.

**Comment by iUltimateLP** [comment]
Thought that also. How do you manage which block is water? Again a hardcoded material index?

**Comment by miguelemosreverte** [comment]
yup. bad habit of mine.

**Comment by miguelemosreverte** [comment]
fixed the bug, uploading video

**Comment by miguelemosreverte** [comment]
https://www.youtube.com/watch?v=AKlmB8aKIVg&feature=youtu.be

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14119347/1ca70374-f5c4-11e5-8613-dbe437c5c87d.png)

**Comment by iUltimateLP** [comment]
Looks cool, much better like Minecraft actually :joy:

**Comment by iUltimateLP** [comment]
Did this actually evolved out of one water source? Isn't it a bit much? :joy:

**Comment by miguelemosreverte** [comment]
3 block sources, but still, lots of water

**Comment by iUltimateLP** [comment]
Is this bad or good?

**Comment by miguelemosreverte** [comment]
I mean, is not like all of that water came out of ust one single source, so its good. It came from three. 

Hey I am working on the water vertices:
https://www.youtube.com/watch?v=jRzuxa5Bqyk&feature=youtu.be

Main problem: How to control them individually! (Probably c++ would do the trick)

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14153877/1514023e-f690-11e5-9f46-ec7f228a6ed7.png)

**Comment by iUltimateLP** [comment]
Does this work as intended?

**Comment by miguelemosreverte** [comment]
No! WIP!

**Comment by miguelemosreverte** [comment]
If you take a look at the materials created by Scheidecker, they all hack FVertexColor, I believe, and use its alpha to send the uint8 value of the AmbientOcclussion. We could copy that so that the C++ info is used by WorldPositionOffset.

**Comment by miguelemosreverte** [comment]
getting there

![image](https://cloud.githubusercontent.com/assets/9152392/14186879/e9db5708-f756-11e5-8ccb-443d6140aed1.png)

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14186905/04e07bbe-f757-11e5-87ab-e8d5534e4361.png)

**Comment by miguelemosreverte** [comment]
Fixed the bug where the lateral faces turn:

replaced this`
SceneProxy->VertexBuffer.Vertices[(LocalVertexCoordinates.Y * LocalVertexDim.X + LocalVertexCoordinates.X) * LocalVertexDim.Z + LocalVertexCoordinates.Z].AmbientOcclusionFactor = 0;`
												
for this `SceneProxy->VertexBuffer.Vertices[VertexIndexMap[(LocalVertexCoordinates.Y * LocalVertexDim.X + LocalVertexCoordinates.X) * LocalVertexDim.Z + LocalVertexCoordinates.Z]].AmbientOcclusionFactor = 0;`
												
[Result](https://www.youtube.com/watch?v=SnSagYfp1g8&feature=youtu.be)

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14194455/277b4298-f785-11e5-9e37-1cc82df57069.png)

**Comment by iUltimateLP** [comment]
That looks cool! Is this depending on how much water the block before has? So if you spill water on the ground, the last one is really flat?

**Comment by miguelemosreverte** [comment]
Not yet, but as you can see, we are getting there!

**Comment by miguelemosreverte** [comment]
![image](https://cloud.githubusercontent.com/assets/9152392/14208300/19bfd680-f7f5-11e5-8c49-59c56c29cf8a.png)

**Comment by miguelemosreverte** [comment]
It would end up looking like this: 
![image](https://cloud.githubusercontent.com/assets/9152392/14208451/cdfb12a4-f7f5-11e5-99ab-693249e0ce8f.png)

**Comment by iUltimateLP** [comment]
Wrong picture? :D

**Comment by iUltimateLP** [comment]
Yeah now it is right, first of you had two times the same picture inside :) Keep me in contact!

**Comment by miguelemosreverte** [comment]
hey with BFS implemented and WorldPositionOffset modifying the top faces of the water bricks I think you have a nice starting point to work towards water simulation. 

I must start spending more time studying so I am going to let you continue on your own.

Some tips:

.By using AmbientOcclusionFactor now the bricks that surround water are really dark, thats a bug that requires fixing
.To continue flooding after well, flooding stops, you are going to have to implement something else. I would suggest going over to the Minecraft implementation of water and checking how they do it.

**Comment by iUltimateLP** [comment]
I think they just check how many iterations are already done and have a limit there.. I only would need to check if it just expands horizontally..

**Comment by miguelemosreverte** [comment]
After all this time I finnally made a somewhat cleaner version of the implementation of the database, both  the one of the BrickContents and the one about the precomputed flooding of the water.
Here's the link:

**Comment by miguelemosreverte** [comment]
How's your implementation of water going?
